oauthproxy.go

```diff
@@ -838,6 +839,44 @@ func (p *OAuthProxy) enrichSessionState(ctx context.Context, s *sessionsapi.Sess
 func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 	session, err := p.getAuthenticatedSession(rw, req)
 	if err != nil {
+		if authOnlyRedirectSignIn(req) {
+			logger.Printf("Initiating AuthOnly SignIn Redirect")
+			if isAjax(req) {
+				// no point redirecting an AJAX request
+				p.errorJSON(rw, http.StatusUnauthorized)
+				return
+			}
+
+			// Option-1: request response is rendered
+			// No duplication of code to get App redirects and relies on SignIn method
+			// p.SignIn(rw, req)
+
+			// Option-2: request response is rendered, but code from `SignIn` method is partially duplicated
+			// if p.SkipProviderButton {
+			// 	p.OAuthStart(rw, req)
+			// } else {
+			// 	p.SignInPage(rw, req, http.StatusForbidden)
+			// }
+
+			// Option-3: Uses HTTP redirect instead of calling SignIn/OauthStart methods
+			// Orignal request URI is captured and added using `rd` parameter in the redirect
+			// to ensure a redirect to the original URL will still happen after OAuth/Signin process
+			redirect, err := p.getAppRedirect(req)
+			if err != nil {
+				logger.Errorf("Error obtaining redirect: %v", err)
+				p.ErrorPage(rw, http.StatusInternalServerError, "Internal Server Error", err.Error())
+				return
+			}
+			logger.Printf("redirect: %s", redirect)
+
+			redirectPath := fmt.Sprintf("%s?rd=%s", p.SignOutPath, redirect)
+			if p.SkipProviderButton {
+				redirectPath = fmt.Sprintf("%s?rd=%s", p.OAuthStartPath, redirect)
+			}
```

Comment:
This will be handled by the sign in page, so I don't think we need to duplicate the logic here

Reply:
Perhaps I'm misunderstanding, but the request flow would be
- https://a-service.example/com/testing
- Traefik ForwardAuth sends to https://oauth2-proxy.example.com/oauth2/auth?redirect_signin=true
- oauth2-proxy checks the session is unauthenticated and redirects back to request URL which will be https://oauth2-proxy.example.com/oauth2/(sign_in|start).

So without the duplication logic of retrieving the original request (https://service-a.example.com) and placing it in an `rd` QS param, the original request is lost

Reply:
Have removed the `SkipProviderButton` duplicated check now, as you pointed out the `SignIn` method is already doing this